docs/source/features/plugins.md

```diff
@@ -0,0 +1,333 @@
+
+For example, to implement any of the request lifecycle events, an object should be returned from `requestDidStart` as such:
+
+```js
+const server = new ApolloServer({
+  /* ... other necessary configuration ... */
+
+  plugins: [
+    {
+      requestDidStart(requestContext) {
+        /* Plugin-specific request scope can be created here and
+           used in more granular lifecycle events below. */
+        return {
+
+          parsingDidStart(requestContext) {
+
+          },
+
+          validationDidStart(requestContext) {
+
+          },
+
+          didResolveOperation(requestContext) {
+
+          },
+
+          /* ... any additional request lifecycle events... */
+
+        }
+      }
+    }
+  ],
+
+  /* */
+})
+```
+
+### `parsingDidStart`
+
+The `parsingDidStart` request lifecycle event will receive the request context as the first argument.  At this stage, the `document` AST may not be defined since the parsing may not succeed.
+
+#### TypeScript signature
+
+```typescript
+parsingDidStart?(
+  requestContext: GraphQLRequestContext<TContext>,
+): (err?: Error) => void | void;
+```
+
+### `validationDidStart`
+
+The `validationDidStart` request lifecycle event will receive the request context as the first argument.  Since parsing would have been successful prior to validation, the `document` AST will be present.
+
+#### TypeScript signature
+
+```typescript
+validationDidStart?(
+  requestContext: WithRequired<GraphQLRequestContext<TContext>, 'document'>,
+): (err?: ReadonlyArray<Error>) => void | void;
+```
+
+### `didResolveOperation`
+
+The `didResolveOperation` request lifecycle event is triggered after the operation to be executed has been successfully retrieved from the `document` AST by `graphql`'s `getOperationAST`.  This focusing of execution which identifies the correct operation is important when a `document` contains multiple operations.  At this stage, in addition to the `document` the `operationName` (`String`) and `operation` (AST) will be present on the context.
+
+If the operation is anonymous (e.g. the operation is `query { ... }` rather than `query NamedQuery { ... }`, then `operationName` will be `null`.
+
+```typescript
+didResolveOperation?(
+  requestContext: WithRequired<
+    GraphQLRequestContext<TContext>,
+    'document' | 'operationName' | 'operation'
+  >,
+): ValueOrPromise<void>;
+```
+
+### `didEncounterErrors`
+
+> TODO
+
+```typescript
+didEncounterErrors?(
+  requestContext: WithRequired<
+    GraphQLRequestContext<TContext>,
+    'metrics' | 'source' | 'errors'
+  >,
+): ValueOrPromise<void>;
+```
+
+### `executionDidStart`
+
+> TODO
+
+```typescript
+executionDidStart?(
+  requestContext: WithRequired<
+    GraphQLRequestContext<TContext>,
+    'document' | 'operationName' | 'operation'
+  >,
+): (err?: Error) => void | void;
+```
+
+### `willSendResponse`
+
+> TODO
+
+```typescript
+willSendResponse?(
```

Comment:
can this be used to modify the response before returning, such as adding additional metrics?

Reply:
Yes!

Reply:
Thanks for including this. If we wanted to alter the extensions object in the response, would either/both of these approaches work?

```
// Approach 1
const helloPlugin1 = {

  requestDidStart() {

    return {
      willSendResponse(requestContext) {

        requestContext.response.extensions.randomField = 'Hello World!'

      },
    };
  },
};

// Approach 2
const _ = require('lodash')

const helloPlugin2 = {

  requestDidStart() {

    return {
      willSendResponse(requestContext) {

        const newReqContext = _.cloneDeep(requestContext);

        newReqContext.response.extensions.randomField = 'Hello World!'

        return newReqContext
      },
    };
  },
};

```